FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+    full batches to submit - including when we're waiting for an auto-flush timer
+    to elapse (it will have to flush again and re-acquire the `Monitor` if someone
+    else hasn't beaten them to it).
+
+- Validate
+    - Ensure we have the right set of event handlers
+    - Ensure we have a key field accessor
+    - This needs to happen on the calling thread so we can raise exceptions
+
+- Adding actions
+    - Call Validate
+    - The `AddActions`/`Async` methods will add everything to the pending
+    actions queue.  It will create an object that includes the action, the
+    action's key, the serialized JSON for the action, and a retry count.  We
+    could consider tracking intermediate failures if we want to expose those in
+    the future.  We'll also `Interlocked.Increment` an `int _pendingSize` with
+    the size of serialized JSON that we can use for deciding when a batch is
+    ready.
+
+    - Validate the event handlers, raise the `ActionAdded`/`Async` events, and
+    log the number of actions added
+
+    - If `AutoFlush` is `true`
+        - Call `TrySubmit` if the batch is full.  If `TrySubmit` returns `false`,
+        we'll just return because someone else is submitting.  If `TrySubmit` is
+        `true` and there's pending actions remaining we'll start the timer.
+
+- Flushing
+    - Call Validate
+    - The `Flush`/`Async` methods will call `TrySubmit(flush: true)`.  If that returns `false`
+    because another thread is already submitting, it will set a `_pendingFlush`
+    flag to `true`.  After the submission is finished, we'd normally check
+    whether another batch was ready to go.  If `_pendingFlush` is true, we'll
+    submit any remaining actions even if they're not a full batch.
+
+- Disposing
+    - Call Validate
+    - The `Dispose`/`DisposeAsync` methods will call `TrySubmit(block: true)` that
+    block on submission.  We'll also set a `_disposed` flag on the publisher.
+    Make sure calling `Dispose` from every sender thread works safely.
+
+- `Task<bool> TrySubmitAsync(bool flush, bool block)`
+    - First it will make sure it's the only thread currently submitting.  It will
+    use `Monitor.TryEnter` when `block == false` and `Monitor.Enter` otherwise.
+        - If we can't acquire the monitor, we'll return `false`.
+        - If we do acquire the monitor, we'll check if we've been `_disposed` and
+        return `false` (in case multiple `Dispose` calls were all waiting on the
+        monitor).
+
+    - Disable any existing timer
+
+    - If `flush || (_pendingSize + _retrySize >= BatchSize)`
+        - Fill the batch
+        - `SubmitBatchAsync(batch, retryAttempt: 0)`
+        - Set `_pendingFlush` to `false`
+
+    - Call `Monitor.Exit` and return `true`
+
+- `Task SubmitAsync(Batch<T> batch, int retryAttempt)`
+    - Raise the ActionSent event for every item in batch
+    - Send the filled batch as a request using our `SearchClient`.
```

Comment:
Should we distinguish between `ActionSent` and `ActionRetried` with an extra event?  So Sent is raised once per document?

Reply:
Could we also put the RetryCount on the ActionSent event?

Reply:
Should we also add an EventArgs class?